libqtile/core/manager.py

```diff
@@ -386,10 +386,18 @@ def grab_chord(self, chord) -> None:
             self.grab_key(key)
 
     def cmd_ungrab_chord(self) -> None:
+        """Leave a chord mode"""
         self.current_chord = False
         hook.fire("leave_chord")
 
         self.ungrab_keys()
+        if self.chord_stack:
+            # The first pop is necessary: Otherwise we would be stuck in a mode;
+            # we could not leave it: the code below would re-enter the old mode.
+            self.chord_stack.pop()
+        else:
+            logger.debug("cmd_ungrab_chord was called when no chord mode was active")
```

Comment:
I think it's clear enough why it'd be needed -- you're just removing the old (current) mode, right? If so then you'd be fine with just:
```py
if not self.chord_stack:
    logger.debug("cmd_ungrab_chord was called when no chord mode was active")
    return
self.chord_stack.pop()
```
You also don't return following the debug log, and continue to `grab_key` the config keys. Is this necessary? Wouldn't the config keys already be grabbed at this point always?

I think it'd be nice if you could merge these 2 commits (the first two, implementation and bug fix for it).

Reply:
Okay, didn't think about that. I should definitely rewrite that.
Regarding the commits: Isn't that called squashing or something? I need to find out how that works, still relatively new to using git. I'd also like to rename the branch since the `persistent` idea was dropped...
One general question: Should prs be always be just a single commit? I came across some merged prs from tych0, ramnes, you etc. so is there a silent agreement that prs should be one commit only?

Reply:
Although PRs are often a single commit, there are also plenty where multiple commits make more sense. Each commit should represent one logical change to the tree and (ideally) the tree would make sense at each point. The way you've committed everything is actually super nice and friendly, and by having more small commits where each one is one change (like you've done) it makes the PR very easy to follow -- so thanks for that!

The git-related changes that I have asked for are simply due to you changing something that you had already changed in a previous commit (e.g. adding `cmd_leave_chord_mode` then in a later commit renaming it). This results in the upstream git history remaining more clean.

You can merge and reorder commits using `git rebase --interactive`, this article might be useful: http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html. It doesn't talk about re-ordering but if you need to do that you can just rearrange the lines in the interactive git file that it opens in an editor.

Reply:
@m-col Okay, thanks a lot for the explanation. I guess you'll like the history rewrite then :smile:
I played through https://learngitbranching.js.org/ so I've heard of the powers of `rebase -i`. But with great power... So naturally I was a bit scared at first (making backup branches just to be sure). But some tests have passed, maybe I'll loose my fear. Marking this as resolved.